### PR TITLE
test: public router unit tests (#292)

### LINKED ISSUE #292: Unit Tests — public.router.ts
## Description
Implement or extend Jest/Supertest tests for `harmony-backend/src/routes/public.router.ts` based on the public router test specification.

## Requirements
- Test file: `harmony-backend/tests/public.router.test.ts`
- Mock Prisma, cache, and rate-limiter dependencies as needed
- Verify no duplicate or overlapping test cases
- Cover all public REST endpoints used by the guest public channel view flow
- Include coverage for guest visibility semantics, especially `PUBLIC_NO_INDEX` vs `PRIVATE`
- Run tests locally and capture coverage
- Resolve or document any spec/implementation mismatch found during testing

## Acceptance Criteria
- [ ] All handlers from the spec have at least one test
- [ ] No hallucinated test cases for non-existent functionality
- [ ] No duplicate or significantly overlapping tests
- [ ] All tests pass locally
- [ ] 80%+ code coverage for `public.router.ts`
- [ ] Tests are isolated — no real DB/Redis/network dependencies required

**Sprint:** 3 (P5 Testing)
**Depends on:** #291
**Context:** Needed to cover the implemented Guest Public Channel View user story and align public-route behavior with the P5 assignment deliverables.

### PR DESCRIPTION
**Pull request by FardeenI**
## Summary

- Extends `harmony-backend/tests/public.router.test.ts` from 16 → 41 tests
- Achieves **100% statement, branch, function, and line coverage** of `public.router.ts`
- Implements all handlers described in the spec from #291
- Regenerates Prisma client (was stale — missing `MessageReaction` model, blocked compilation)

## What was added

| Route | New tests |
|---|---|
| `GET /servers` | ordering, 20-cap, empty result, `isPublic` filter, 500 |
| `GET /servers/:serverSlug` | `X-Cache: HIT`, `X-Cache: STALE`, Redis error, 500 on `getOrRevalidate` |
| `GET /servers/:serverSlug/channels` | `orderBy` assertion, HIT/STALE cache headers, 500 |
| `GET /servers/:serverSlug/channels/:channelSlug` | `PUBLIC_INDEXABLE` 200, `PUBLIC_NO_INDEX` 200, `PRIVATE` 403, server 404, channel 404, scoped lookup assertion, 500 |
| `GET /channels/:channelId/messages` | `isDeleted: false` filter assertion, 500 |
| `GET /channels/:channelId/messages/:messageId` | 500 |
| Rate limiting | pass-through (PR-22), 429 exhaustion (PR-23) |

## Mock changes

- Added `server.findMany` and `channel.findFirst` to the Prisma mock factory
- Imported `cacheService` for per-test `mockResolvedValueOnce` overrides on HIT/STALE/error paths

## Base branch

This PR targets `test/issue-291-public-router-spec` (the spec PR #302). It should be merged after that PR lands on main.

## Test plan

- [x] `npx jest tests/public.router.test.ts` — 41/41 passing
- [x] 100% coverage across statements, branches, functions, lines

Closes #292

### CODE REVIEW

**Review by acabrera04 — CHANGES_REQUESTED** [review]
- [ ] 1. Prioritize Logic over Style: unresolved reliability/install concerns remain, so this is not a clean pass yet.
- [x] 2. Security First: this PR only expands tests and does not introduce new user-facing attack surface.
- [x] 3. Architectural Alignment: the added cases match the current `public.router.ts` behavior and public-route architecture.
- [x] 4. Issue Completion: the PR covers the linked issue’s required handlers, guest-visibility semantics, and isolated mocking strategy.
- [x] 5. No Nitpicking: the only remaining concerns are substantive, not stylistic.
- [x] 6. Avoid Repetition: I checked the existing review comments and am not adding duplicate inline feedback.
- [x] 7. Iterative Reviews: existing comments were reviewed; there are no follow-up commits addressing them yet.
- [ ] 8. Prevent CI Failures: the already-raised rate-limit test flake and lockfile churn are still plausible CI/install risks.

Requesting changes. I did not find any additional issues beyond the two unresolved inline comments already on the PR: the timing-dependent token-bucket exhaustion test in `harmony-backend/tests/public.router.test.ts` and the unintended-looking `fsevents` lockfile hunk in `harmony-backend/package-lock.json`. After those are addressed, the rest of the test expansion looks sound and aligned with issue #292.

Verification I ran locally on the PR branch in a temporary worktree:
- `npx jest tests/public.router.test.ts --runInBand` -> 41/41 passing
- `npx eslint tests/public.router.test.ts` -> passed

File: harmony-backend/tests/public.router.test.ts [hunk 1]
```diff
@@ -334,4 +341,325 @@ describe('GET /api/public/channels/:channelId/messages/:messageId', () => {
...
+    mockPrisma.server.findMany.mockResolvedValue([]);
+
+    // Exhaust the 100-token human bucket
+    for (let i = 0; i < 100; i++) {
+      await request(app).get('/api/public/servers');
+    }
+
+    const res = await request(app).get('/api/public/servers');
+    expect(res.status).toBe(429);
+    expect(res.body).toHaveProperty('error');
```
[the remark this answers is not in the document]
Comment (FardeenI) [reply harmony-backend/tests/public.router.test.ts]: Fixed in 0d82cf0. Froze `Date.now` via `jest.spyOn` in a try/finally block so the token bucket cannot partially refill between requests — the 101st call is now deterministically 429 on any CI speed.

**Comment by FardeenI** [comment]
> Re: comment 4058962206

Both inline concerns are now addressed in 0d82cf0: the rate-limit test is frozen with `jest.spyOn(Date, 'now')` to prevent token-bucket refill flake, and the `fsevents` lockfile entry has its `dev: true` marker restored. 41/41 tests passing.

**Review by acabrera04 — CHANGES_REQUESTED** [review]
- [ ] 1. Prioritize Logic over Style: the earlier flake/lockfile issues are fixed, but the test suite still does not implement several required cases from the updated spec in PR #302.
- [x] 2. Security First: this PR is still test-only and does not introduce new runtime attack surface.
- [x] 3. Architectural Alignment: the assertions that are present still match the current `public.router.ts` architecture.
- [ ] 4. Issue Completion: against the latest `docs/test-specs/public-router-spec.md` from PR #302, coverage is still incomplete.
- [x] 5. No Nitpicking: the remaining concerns are about missing required test coverage, not style.
- [x] 6. Avoid Repetition: I am not repeating the prior rate-limit/lockfile feedback because commit `0d82cf0` addressed those.
- [x] 7. Iterative Reviews: I checked the follow-up commit and the previous findings are resolved.
- [ ] 8. Prevent CI Failures: the suite passes, but it still leaves several spec-mandated branches untested, so regressions could still slip through while CI stays green.

The previous findings are fixed, but I still can’t approve this against the updated spec in PR #302.

1. `docs/test-specs/public-router-spec.md` §4.3 now requires the cache-middleware cases PR-17 through PR-21b, including HIT/STALE/MISS behavior and the distinct single-message cache key. `harmony-backend/tests/public.router.test.ts` still jumps from the route-2 handler assertions straight into `/servers` coverage without ever asserting `cacheMiddleware` behavior for routes 1 and 2, so the most stateful part of the public router is still unverified.
2. `docs/test-specs/public-router-spec.md` §4.5 requires PR-29 and PR-33 for `/servers/:serverSlug`: a cache-MISS response that asserts `X-Cache-Key`/MISS wiring, and a 404 case that proves `cacheService.getOrRevalidate` is not called. The current `/servers/:serverSlug` tests only cover basic 200/404 plus HIT/STALE/Redis-error/500 paths, so those two spec cases are still missing.
3. The updated spec still calls for explicit pagination/message-edge cases that are not covered here: PR-5 (non-numeric `?page=abc` defaults to page 1) and PR-15 (single-message lookup is filtered by `isDeleted: false`, not just a generic `null` result). Right now the suite checks `page=0`/negative and a missing message, but it never proves those two branches from the spec.

Verification I ran on the current branch:
- `npx jest tests/public.router.test.ts --runInBand` -> 41/41 passing
- `npx eslint tests/public.router.test.ts` -> passed

**Review by acabrera04 — CHANGES_REQUESTED** [review]
- [ ] 1. Prioritize Logic over Style: the branch is much closer, but two core behavior checks from the updated spec are still not actually asserted.
- [x] 2. Security First: this remains test-only work with no new runtime attack surface.
- [x] 3. Architectural Alignment: the test structure still matches the current `public.router.ts` and cache middleware architecture.
- [ ] 4. Issue Completion: against the latest spec in PR #302, coverage is still not fully aligned on pagination and stale-revalidation behavior.
- [x] 5. No Nitpicking: the remaining feedback is about missing behavioral assertions, not style.
- [x] 6. Avoid Repetition: the previous flake/lockfile/spec-gap feedback is resolved in `0d82cf0` and `3f02476`, so I am not repeating it.
- [x] 7. Iterative Reviews: I checked the new commit and the earlier requested changes were addressed.
- [ ] 8. Prevent CI Failures: the suite passes, but these two gaps still allow real regressions to slip through while CI stays green.

The new commit fixes the previous review points, but I still see two blocking gaps relative to the updated spec in PR #302:

1. The pagination tests for `GET /api/public/channels/:channelId/messages` still only assert the response body, not the Prisma pagination call shape that the spec requires. In `docs/test-specs/public-router-spec.md` §4.1, PR-2 and PR-3 explicitly require asserting `skip: 100, take: 50` for `?page=3` and `skip: 0` for `?page=0`. In `harmony-backend/tests/public.router.test.ts`, the `?page=3` and clamping cases still stop at `page` in the response, so a regression in the `skip`/`take` wiring would still pass these tests.
2. The new PR-18 stale-cache test proves the stale response body is served, but it still does not prove the background revalidation path actually runs. The updated spec’s §4.3 definition of PR-18 requires the stale response plus handler execution in the background (`Prisma is called`). Right now the test asserts `X-Cache: STALE` and the stale body, but never verifies that revalidation reached the Prisma layer, so a regression where stale entries are served without revalidation would still satisfy the test.

Verification I ran on the current head:
- `npx jest tests/public.router.test.ts --runInBand` -> 51/51 passing
- `npx jest tests/public.router.test.ts --runInBand --coverage --collectCoverageFrom=src/routes/public.router.ts` -> 51/51 passing
- `npx eslint tests/public.router.test.ts` -> passed

**Review by acabrera04 — CHANGES_REQUESTED** [review]
- [ ] 1. Prioritize Logic over Style: the latest commit fixed the prior issues, but one required route-behavior case from the updated spec is still missing.
- [x] 2. Security First: this is still test-only work with no runtime security impact.
- [x] 3. Architectural Alignment: the suite structure matches the current router and cache middleware design.
- [ ] 4. Issue Completion: the updated spec in PR #302 still is not fully covered.
- [x] 5. No Nitpicking: the remaining feedback is about a missing behavioral test, not style.
- [x] 6. Avoid Repetition: I am not repeating the earlier flake, lockfile, pagination, or stale-revalidation findings because those are now addressed.
- [x] 7. Iterative Reviews: I checked the latest commit and confirmed the previous review comments were fixed.
- [ ] 8. Prevent CI Failures: the suite passes, but this missing MISS-path assertion still leaves one public-route cache branch unprotected.

The latest head fixes the previous review points, but I still see one remaining spec gap.

1. `docs/test-specs/public-router-spec.md` §4.6 still requires PR-39 for `GET /servers/:serverSlug/channels`: when `cacheService.get` throws, the route should continue with `X-Cache: MISS` and return 200. In `harmony-backend/tests/public.router.test.ts`, the `/servers/:serverSlug/channels` block covers ordering, HIT, STALE, and 500 (`PR-35`, `PR-39b`, `PR-39c`, `PR-40`), but it still never exercises the Redis-error MISS path for this route. That means a regression in the silent-catch fallback for server-channel caching would still pass the suite.

Verification I ran on the current head:
- `npx jest tests/public.router.test.ts --runInBand` -> 51/51 passing
- `npx jest tests/public.router.test.ts --runInBand --coverage --collectCoverageFrom=src/routes/public.router.ts` -> 51/51 passing
- `npx eslint tests/public.router.test.ts` -> passed

**Review by acabrera04 — CHANGES_REQUESTED** [review]
- [ ] 1. Prioritize Logic over Style: the remaining gap is small, but it still leaves spec-required response-header behavior untested.
- [x] 2. Security First: this is test-only work with no new runtime security surface.
- [x] 3. Architectural Alignment: the suite still follows the current router and cache-middleware architecture.
- [ ] 4. Issue Completion: against the updated spec in PR #302, header coverage is still incomplete for two route families.
- [x] 5. No Nitpicking: the remaining feedback is about missing behavioral assertions, not style.
- [x] 6. Avoid Repetition: the earlier rate-limit, lockfile, cache, pagination, and Redis-error gaps are now fixed, so I am not repeating them.
- [x] 7. Iterative Reviews: I checked the latest commits and confirmed the prior findings were addressed.
- [ ] 8. Prevent CI Failures: the suite passes, but these header regressions could still slip through while CI stays green.

The latest head fixed the previous findings, but I still see one remaining spec-alignment issue.

1. The updated spec still requires header assertions that are missing for the two server-channel endpoints. In `docs/test-specs/public-router-spec.md` §4.6, PR-35 explicitly includes `Cache-Control: public, max-age=300` and `X-Cache-Key` for `GET /servers/:serverSlug/channels`, but the tests for that route only cover body contents, ordering, X-Cache HIT/STALE/MISS, and the 500 path. Likewise, §4.7 PR-41 expects `Cache-Control: public, max-age=300` on `GET /servers/:serverSlug/channels/:channelSlug`, and the current test only checks status/body. Those headers are part of the public caching contract in `public.router.ts`, so they should be asserted directly.

Verification I ran on the current head:
- `npx jest tests/public.router.test.ts --runInBand` -> 52/52 passing
- `npx jest tests/public.router.test.ts --runInBand --coverage --collectCoverageFrom=src/routes/public.router.ts` -> 52/52 passing
- `npx eslint tests/public.router.test.ts` -> passed

**Comment by FardeenI** [comment]
> Re: comment 4059039172

Already addressed in 7cf3a68. Added cache middleware tests PR-17 through PR-21b, PR-29/PR-33, and PR-5/PR-15.

**Comment by FardeenI** [comment]
> Re: comment 4058962206

Already addressed in d30d4f5. Fixed rate-limit test flake via Date.now spy and resolved lockfile churn.

**Review by acabrera04 — APPROVED** [review]
- [x] 1. Prioritize Logic over Style: the remaining behavioral gaps from earlier reviews are now covered with concrete assertions.
- [x] 2. Security First: this PR is test-only and introduces no new runtime security surface.
- [x] 3. Architectural Alignment: the suite matches the current `public.router.ts` and cache middleware behavior.
- [x] 4. Issue Completion: against the updated spec in PR #302, the required route and header cases are now covered.
- [x] 5. No Nitpicking: no remaining feedback beyond substantive behavior checks.
- [x] 6. Avoid Repetition: prior findings were rechecked and are now addressed.
- [x] 7. Iterative Reviews: I verified the latest head commit `69d9147` and confirmed the previously requested changes were implemented.
- [x] 8. Prevent CI Failures: the targeted test file and linting both pass on the latest PR head.

No remaining findings from my side. The latest push adds the missing `PR-35` and `PR-41` header assertions, and the earlier review gaps for rate-limit determinism, cache-middleware behavior, pagination `skip`/`take`, stale revalidation, and the `/servers/:serverSlug/channels` Redis-error MISS path are all covered now.

Verification I ran on the latest PR head (`69d9147c967c899056acc9c71a02edf1aa9646fb`):
- `npx jest tests/public.router.test.ts --runInBand` -> 52/52 passing
- `npx eslint tests/public.router.test.ts` -> passed
